### DISCUSSION TITLE
generate volumemout jsonpatch array depending on if container already have volumemount

### DISCUSSION THREAD

**Discussion by zhoupoko2000** [open-policy-agent/community#87] [OPA and Rego] [answered]
Hello Team,
I am trying to achieve below goal. so for input like this
`{
    "request": {"spec": "haha"},
    "message": "make",
    "containers": [{"name": "app1"}, {"name": "app2"}, {"name": "app3", "volmount": [{"name": "volmount_al"}]}]
}`

I want to have a jsonpatch like below, basically, generate new attribute of volmount if container does not have this attribute yet.
However, append new volmount1 to existing volmount if certain container already has this attribute.
Here, the number of container and which container already has attribute volmount can be dynamic/random here.

`[{"op": "add", "path": "/containers/0/volmount", "value": [{"name": "volmount1"}]}, {"op": "add", "path": "/containers/1/volmount", "value": [{"name": "volmount1"}]}, {"op": "add", "path": "/containers/2/volmount/-", "value": [{"name": "volmount1"}]}]`

basically, this stuck as the if else rego block just CANNOT be embedded into any rule or function block.
so far, this is what I get

> package play

make_vol_patch(path, num_str, value) = patchCode {
    patchCode := {
        "op": "add",
        "path": concat("/", [path, num_str, "volmount"]),
        "value": [value]
    }
}

add_vol_patch(path, num_str, value) = patchCode {
	patchCode := {
    	"op": "add",
        "path": concat("/", [path, num_str, "volmount", "-"]),
        "value": [value]

    }
}
has_volm(obj) {
	obj.volmount
}

vm_value := {"name": "volmount1"}

patches[patch]{

    some i
    container_i := input.containers[i]
    i_str := format_int(i, 10)

    patch = add_vol_patch("/containers", i_str, vm_value) {
        has_volm(container_i)
    } else = make_vol_patch("/containers", i_str, vm_value)

}

**Comment by anderseknert** [comment, marked as answer]
The function should not be defined inside of the rule, but can definitely be called from there. Something like this should work:

```rego
package play

vm_value := {"name": "volmount1"}

patches[patch] {
	some i
	container_i := input.containers[i]
	patch := {
		"op": "add",
		"path": concat("/", get_path(i, object.get(container_i, "volmount", false))),
		"value": [vm_value],
	}
}

get_path(i, exists) = ["/containers", format_int(i, 10), "volmount", "-"] {
	exists
}

get_path(i, exists) = ["/containers", format_int(i, 10), "volmount"] {
	not exists
}

```

Though if you are trying to build the path dynamically for mutation, that's actually trickier than it seems. Have a look at [this repo](https://github.com/tsandall/validating-and-mutating-example) from @tsandall for some example code doing what I _think_ you are trying to do :)

**Reply by zhoupoko2000** [reply]
Thanks! @anderseknert  this is so helpful.
